bochs/bx_debug/dbg_main.cc

```diff
@@ -1939,6 +1987,119 @@ void bx_dbg_deref_command(bx_address addr, unsigned deep)
   }
 }
 
+void bx_dbg_telescope(bx_address addr)
+{
+  unsigned deep_sc = 0;
+  bx_address deref_data = 0;
+  bx_address last_valid_addr = 0;
+  bx_address curr_addr = 0;
+  unsigned char data[MAX_TELESCOPE_DATA_DEPTH + 1];
+  bool print_data = false;
+  
+  if (telescope_max_depth == 0 || telescope_max_depth >= MAX_TELESCOPE_DATA_DEPTH)
+    return;
+
+  last_valid_addr = addr;
+  curr_addr = addr;
+  for (int i = 0; i < telescope_max_depth + 1; i++) {
+    deref_data = bx_dbg_deref(curr_addr, 1, &deep_sc, NULL, true);
+    if (0 != deep_sc) {
+      print_data = true;
+      break;
+    }
+
+    if (i != 0)
+      dbg_printf("0x" FMT_ADDRX " ", curr_addr);
+
+    last_valid_addr = curr_addr;
+
+    if (deref_data == curr_addr) {
+      dbg_printf("iloop");
+      break;
+    }
+
+    curr_addr = deref_data;  
+  }
+
+  if (print_data && bx_dbg_read_linear(dbg_cpu, last_valid_addr, telescope_data_depth, data, true)) {
+    if (telescope_data_hex) {
+      dbg_printf(" h'");
+      for (int i = 0; i < telescope_data_depth; i++) {
+        dbg_printf("%02x", data[i]);
+      }
+      dbg_printf("'");
+    }
+    else {
+      dbg_printf(" s'");
+      for (int i = 0; i < telescope_data_depth; i++) {
+        if (data[i] < 0x20 || data[i] > 0x7E)
+          data[i] = '.';
+      }
+      data[telescope_data_depth] = '\0';
+      dbg_printf("%s'", data);
+    }
+  }
+}
+
+void bx_dbg_telescopeset_command(unsigned int max_rows, unsigned int max_depth, unsigned int data_depth, bool data_hex)
+{
+  if (data_depth > MAX_TELESCOPE_DATA_DEPTH)
+    telescope_data_depth = MAX_TELESCOPE_DATA_DEPTH;
+  else
+    telescope_data_depth = data_depth;
+
+  telescope_max_rows = max_rows;
+  telescope_max_depth = max_depth;
+  telescope_data_hex = data_hex;
+}
+
+void bx_dbg_telescope_command(bx_address addr)
+{
+  unsigned len;
+
+  BX_CPU_C *cpu = BX_CPU(dbg_cpu);
+
+#if BX_SUPPORT_X86_64
+  if (cpu->get_cpu_mode() == BX_MODE_LONG_64) {
+    len = 8;
+  }
+  else
+#endif
+  {
+    if (cpu->sregs[BX_SEG_REG_SS].cache.u.segment.d_b) {
+      len = 4;
+    }
+    else {
+      len = 2;
+    }
+  }
+
+  for (unsigned i = 0; i < telescope_max_rows; i++) {
+#if BX_SUPPORT_X86_64
+    if (len == 8) {
+      dbg_printf("0x%08x%08x ", GET32H(addr), GET32L(addr));
+      bx_dbg_telescope(addr);
+      dbg_printf("\n");
+    }
+    else
+#endif
+    {
+      if (len == 4) {
+        dbg_printf("0x%08x ", (unsigned) addr);
+        bx_dbg_telescope(addr);
```

Comment:
Avoid duplicating bx_dbg_telescope(addr) call ?
just do dbg_printf with if's and call bx_dbg_telescope(addr) after them